drivers/kscan/kscan_matrix_handlers.c

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright (c) 2022 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/kscan.h>
+#include <syscall_handler.h>
+
+static int z_vrfy_kscan_matrix_configure(const struct device *dev,
+					 kscan_isr_callback_t isr_callback)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_KSCAN(dev, matrix_config));
+
+	return z_impl_kscan_matrix_configure((const struct device *)dev, isr_callback);
+}
+#include <syscalls/kscan_matrix_configure_mrsh.c>
+
+static int z_vrfy_kscan_matrix_drive_column(const struct device *dev, int col)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_KSCAN(dev, matrix_drive_column));
+
+	return z_impl_kscan_matrix_drive_column((const struct device *)dev, col);
+}
+#include <syscalls/kscan_matrix_drive_column_mrsh.c>
+
+static int z_vrfy_kscan_matrix_read_row(const struct device *dev, int *row)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_KSCAN(dev, matrix_read_row));
+
+	return z_impl_kscan_matrix_read_row((const struct device *)dev, row);
```

Comment:
Are you sure about this? It looks to me like a malicious userspace thread could pass an address in `row` that they don't have write access to. But the (privileged mode) driver unconditionally writes to the address pointed by `row` here:

```c
	/* 1 means key pressed, otherwise means key released. */
	*row = (~val & NPCX_KB_ROW_MASK);
```

... this looks like a perfect recipe for an arbitrary write primitive. What am I missing here?

Reply:
In addition to verifying `*dev`, you will also need to verify that `*row` can be readable, or writable, or read/write-able by using `Z_SYSCALL_MEMORY_*`.

Reply:
I added the write permission check for it:
`Z_OOPS(Z_SYSCALL_MEMORY_WRITE(row, sizeof(*row))); `